mender-update/daemon/state_machine.cpp

```diff
@@ -140,9 +146,85 @@ StateMachine::DeploymentTracking::DeploymentTracking() :
+		log::Debug("No existing deployment data, starting from initial state");
+		return;
+	}
+
+	// We have state data, move it to the context.
+	ctx_.deployment.state_data = std::move(state_data);
+
+	auto &state = ctx_.deployment.state_data->state;
+
+	if (state == ctx_.kUpdateStateDownload) {
+		main_states_.SetState(update_cleanup_state_);
+		// "rollback_attempted_state" because Download in its nature makes no system
+		// changes, so a rollback is a no-op.
+		deployment_tracking_.states_.SetState(deployment_tracking_.rollback_attempted_state_);
+
+	} else if (state == ctx_.kUpdateStateArtifactReboot) {
+		// Normal update path with a reboot.
+		main_states_.SetState(update_verify_reboot_state_);
+		deployment_tracking_.states_.SetState(deployment_tracking_.no_failures_state_);
+
+	} else if (state == ctx_.kUpdateStateArtifactRollback) {
+		// Installation failed, but rollback could still succeed.
+		main_states_.SetState(update_rollback_state_);
+		deployment_tracking_.states_.SetState(deployment_tracking_.rollback_attempted_state_);
+
+	} else if (
+		state == ctx_.kUpdateStateArtifactRollbackReboot
+		|| state == ctx_.kUpdateStateArtifactVerifyRollbackReboot
+		|| state == ctx_.kUpdateStateVerifyRollbackReboot) {
+		// Normal flow for a rebooting rollback.
+		main_states_.SetState(update_verify_rollback_reboot_state_);
+		deployment_tracking_.states_.SetState(deployment_tracking_.rollback_attempted_state_);
+
+	} else if (
+		state == ctx_.kUpdateStateAfterArtifactCommit
+		|| state == ctx_.kUpdateStateUpdateAfterFirstCommit) {
+		// Re-run commit Leave scripts if spontaneously rebooted after commit.
+		main_states_.SetState(update_after_commit_state_);
+		deployment_tracking_.states_.SetState(deployment_tracking_.no_failures_state_);
+
+	} else if (state == ctx_.kUpdateStateArtifactFailure) {
+		main_states_.SetState(update_failure_state_);
+		// Actually we don't know if the rollback failed or not; we don't record enough
+		// information in the database to tell. Probably better to err on the side of
+		// caution and report error. The logs should reveal if there really was an error or
+		// not.
+		deployment_tracking_.states_.SetState(deployment_tracking_.failure_state_);
+
+	} else if (state == ctx_.kUpdateStateCleanup) {
+		main_states_.SetState(update_cleanup_state_);
+		// Actually we don't know if the update failed or not; we don't record enough
+		// information in the database to tell. Probably better to err on the side of
+		// caution and report error. The logs should reveal if there really was an error or
+		// not.
+		deployment_tracking_.states_.SetState(deployment_tracking_.failure_state_);
+
+	} else {
+		// All other states trigger a rollback.
+		main_states_.SetState(update_check_rollback_state_);
+		deployment_tracking_.states_.SetState(deployment_tracking_.failure_state_);
+	}
+
+	auto &payload_types = ctx_.deployment.state_data->update_info.artifact.payload_types;
+	assert(payload_types.size() == 1);
```

Comment:
I would be careful with this I think. Why not do a check and log ? This should be handled in the `parser` and nowhere else imio

Reply:
Actually, this is exactly why I added an assert and nothing else. I don't expect this to ever be triggered, because it is handled in the parser. But if we ever add support to the parser, then this will need adapting too, and at that point it's good to have this assert as a reminder.